http/middleware/log_request.go

```diff
@@ -2,44 +2,167 @@ package middleware
-			if query != "" {
-				uri += "?" + q.Encode()
-			}
-
-			strs := []string{r.Method, uri}
-			val := r.Context().Value(IpAddrCtxKey)
-			if val != nil {
-				strs = append([]string{val.(string)}, strs...)
-			}
-
-			ls.Info(strings.Join(strs, " "), nil)
-			h.ServeHTTP(w, r)
+			start := time.Now()
+
+			writer := &requestLogger{ResponseWriter: w, status: http.StatusOK}
+			h.ServeHTTP(writer, r)
+
+			end := time.Since(start).Milliseconds()
+
+			rec := newRecord(writer, r)
+			rec.Duration = end
+
+			var msg string // NOTE(dlk): no message for now.
+
+			// NOTE(dlk): LogAttrs is not the standard method to use,
+			// but we know we only have a []slog.Attr to report,
+			// so let's leverage it.
+			//
+			// TODO(dlk): using r.Context() may not be appropriate.
+			// Unfortunately, outside of configuring a slog.Handler to pull
+			// values from a context.Context into slog.Attr,
+			// contexts appear to not be used by slog.
+			// This includes not watching for cancellations.
+			// The ranger.Context may be the best fit for this.
+			// We'll reassess when slog makes into a Go version.
+			ls.LogAttrs(r.Context(), slog.LevelInfo, msg, rec.attrs()...)
 		})
 	}
 }
+
+// A LogRequestRecord represents the fields that a LogRequest
+type LogRequestRecord struct {
+	BodySize       int    `json:"responseBodySize"`
+	Duration       int64  `json:"duration"`
+	Host           string `json:"host"`
+	ID             string `json:"id"`
+	IPAddr         string `json:"remoteAddr"`
+	Method         string `json:"method"`
+	Path           string `json:"path"`
+	Protocol       string `json:"protocol"`
+	Referrer       string `json:"referrer"`
+	ReqContentLen  int    `json:"requestContentLength"`
+	ReqContentType string `json:"requestContentType"`
+	Scheme         string `json:"scheme"`
+	Status         int    `json:"status"`
+	URI            string `json:"uri"`
+	UserAgent      string `json:"userAgent"`
+}
+
+// newRecord constructs a record from the values availabe in w & r.
+func newRecord(w *requestLogger, r *http.Request) LogRequestRecord {
+	// TODO(dlk): if there's a compelling reason for constructing a LogRequestRecord
+	// outside this package, this constructor and LogRequestRecord.attrs could be exported.
+	uri := new(url.URL)
+	*uri = *r.URL
+	q := r.URL.Query()
+	mask(q, passwordParam)
+	uri.RawQuery = q.Encode()
+
+	contLen, _ := strconv.Atoi(r.Header.Get(contentLenHeader))
+	id, _ := r.Context().Value(trails.RequestIDKey).(string)
+	ip, _ := r.Context().Value(trails.IpAddrKey).(string)
+
+	return LogRequestRecord{
+		BodySize:       w.bodysize,
+		Host:           r.Host,
+		ID:             id,
+		IPAddr:         ip,
+		Method:         r.Method,
+		Path:           r.URL.Path,
+		Protocol:       r.Proto,
+		Referrer:       r.Header.Get(referrerHeader),
+		ReqContentLen:  contLen,
+		ReqContentType: r.Header.Get(contentTypeHeader),
+		Scheme:         r.URL.Scheme,
+		Status:         w.status,
+		URI:            uri.RequestURI(),
```

Comment:
what's the difference between `r.URL.RequestURI()` and `r.RequestURI`, or are they the same?

Reply:
I had a chance to look at the source and it seems they are equivalent: https://cs.opensource.google/go/go/+/refs/tags/go1.20.3:src/net/http/request.go;l=1052

Why wouldn't we favor `r.RequestURI`  here instead to avoid the extra hop into parse and hop back to unparse? What does processing through `net/url` gain us?

Reply:
Sorry, I'm only getting intermittent chances to look into this (which is probably annoying) -- but glancing at the 2nd half of that function linked above makes me wonder if the hop through `net/url` _is_ actually useful to appropriately account for uncommon request cases such as `CONNECT`: https://cs.opensource.google/go/go/+/refs/tags/go1.20.3:src/net/http/request.go;l=1073-1080;bpv=0 

But maybe I just need some helping grokking what's going on here.

Reply:
Yeah the little jig on lines 90-94 is the main reason given some values in the URI are masked.